tests/lib/MVC/Symfony/Security/User/BaseProviderTestCase.php

```diff
@@ -0,0 +1,195 @@
+            ->willReturn($apiUser);
+
+        $this->userService
+            ->expects(self::once())
+            ->method('loadUser')
+            ->with($userId)
+            ->willThrowException(new NotFoundException('user', 'foo'));
+
+        $this->expectException(UserNotFoundException::class);
+        $this->userProvider->refreshUser($user);
+    }
+
+    public function testRefreshUserNotSupported(): void
+    {
+        $user = $this->createMock(SymfonyUserInterface::class);
+
+        $this->expectException(UnsupportedUserException::class);
+        $this->userProvider->refreshUser($user);
+    }
+
+    protected function createUserWrapperMockFromAPIUser(User $apiUser, int $userId): UserInterface & MockObject
+    {
+        $refreshedAPIUser = clone $apiUser;
+        $user = $this->createMock(UserInterface::class);
+        $user
+            ->expects(self::once())
+            ->method('getAPIUser')
+            ->willReturn($apiUser)
+        ;
+        $user
+            ->expects(self::once())
+            ->method('setAPIUser')
+            ->with($refreshedAPIUser)
+        ;
+
+        $this->userService
+            ->expects(self::once())
+            ->method('loadUser')
+            ->with($userId)
+            ->willReturn($refreshedAPIUser)
+        ;
+
+        return $user;
+    }
+
+    public function testRefreshUser(): void
+    {
+        $userId = 123;
+        $apiUser = $this->buildUserValueObjectStub($userId);
+        $user = $this->createUserWrapperMockFromAPIUser($apiUser, $userId);
+
+        $this->permissionResolver
+            ->expects(self::once())
+            ->method('setCurrentUserReference')
+            ->with(new UserReference($apiUser->getUserId()));
+
+        self::assertSame($user, $this->userProvider->refreshUser($user));
+    }
+
+    public function testLoadUserByUsername(): void
+    {
+        $username = $this->getUserIdentifier();
+        $apiUser = $this->createMock(APIUser::class);
+
+        $this->userService
+            ->expects(self::once())
+            ->method($this->getUserServiceMethod())
+            ->with($username)
+            ->willReturn($apiUser);
+
+        $user = $this->userProvider->loadUserByIdentifier($username);
+        self::assertInstanceOf(UserInterface::class, $user);
+        self::assertSame($apiUser, $user->getAPIUser());
+        self::assertSame(['ROLE_USER'], $user->getRoles());
+    }
+
+    public function testLoadUserByUsernameUserNotFound(): void
+    {
+        $username = $this->getUserIdentifier();
+        $this->userService
+            ->expects(self::once())
+            ->method($this->getUserServiceMethod())
+            ->with($username)
+            ->willThrowException(new NotFoundException('user', $username));
+
+        $this->expectException(UserNotFoundException::class);
+        $this->userProvider->loadUserByIdentifier($username);
+    }
+
+    protected function buildUserValueObjectStub(int $userId): User
```

Comment:
Unless you expect it to be changed in descendants.
```suggestion
    final protected function buildUserValueObjectStub(int $userId): User
```